File: src/tacky.ts
/**
 * TACKY IR definition
 */

import { BinaryOpName } from "./ast";
import { Identifier } from "./shared";

export type Program = {
  func: Function;
};
export function Program(func: Function): Program {
  return { func };
}

export type Function = {
  name: Identifier;
  body: Instruction[];
};
export function Function(name: Identifier, body: Instruction[]): Function {
  return { name, body };
}

export type Return = {
  kind: "return";
  src: Value;
};
export function Return(val: Value): Return {
  return { kind: "return", src: val };
}

export type UnaryMinus = {
  kind: "unary-minus";
  src: Value;
  dst: Assignable;
};
export function UnaryMinus(src: Value, dst: Assignable): UnaryMinus {
  return { kind: "unary-minus", src, dst };
}

export type Complement = {
  kind: "complement";
  src: Value;
  dst: Assignable;
};
export function Complement(src: Value, dst: Assignable): Complement {
  return { kind: "complement", src, dst };
}

export type LogicalNot = {
  kind: "logical-not";
  src: Value;
  dst: Assignable;
};
export function LogicalNot(src: Value, dst: Assignable): LogicalNot {
  return { kind: "logical-not", src, dst };
}

export type BinaryOp = {
  kind: "binary-expr";
  operator: BinaryOpName;
  lhs: Value;
  rhs: Value;
  dst: Assignable;
};
export function BinaryOp(
  operator: BinaryOpName,
  lhs: Value,
  rhs: Value,
  dst: Assignable
): BinaryOp {
  return { kind: "binary-expr", operator, lhs, rhs, dst };
}

export type Jump = {
  kind: "jump";
  target: Identifier;
};
export function Jump(target: Identifier): Jump {
  return { kind: "jump", target };
}

export type JumpIfZero = {
  kind: "jump-if-zero";
  condition: Value;
  target: Identifier;
};
export function JumpIfZero(condition: Value, target: Identifier): JumpIfZero {
  return { kind: "jump-if-zero", condition, target };
}

export type JumpIfNotZero = {
  kind: "jump-if-not-zero";
  condition: Value;
  target: Identifier;
};
export function JumpIfNotZero(
  condition: Value,
  target: Identifier
): JumpIfNotZero {
  return { kind: "jump-if-not-zero", condition, target };
}

export type Label = {
  kind: "label";
  name: Identifier;
};
export function Label(name: Identifier): Label {
  return { kind: "label", name };
}

export type Copy = {
  kind: "copy";
  src: Value;
  dst: Assignable;
};
export function Copy(src: Value, dst: Assignable): Copy {
  return { kind: "copy", src, dst };
}

export type UnaryOp = UnaryMinus | Complement | LogicalNot;

export type Instruction =
  | Return
  | UnaryOp
  | BinaryOp
  | Jump
  | JumpIfNotZero
  | JumpIfZero
  | Label
  | Copy;

export type Constant = {
  kind: "numeric-constant";
  val: number;
};
export function Constant(val: number): Constant {
  return { kind: "numeric-constant", val };
}

export type Var = {
  kind: "var";
  name: Identifier;
};
export function Var(name: Identifier): Var {
  return { kind: "var", name };
}

/**
 * Values can store values and include things which cannot be written
 */
export type Value = Constant | Var;

/**
 * Assignable entities can be written to, i.e. be the `dst` of an operation
 */
export type Assignable = Var;

File: rust_src/tacky.rs
//! The definition for the TACKY three address code IR used in the middle end

use crate::{ast, shared_types::Identifier};

/// TACKY Program representation
pub struct Program {
  pub function: Function,
}

/// TACKY Function representation
pub struct Function {
  pub name: Identifier,
  pub body: Vec<Instruction>,
}

/// TACKY Value representation
#[derive(Debug, Clone, PartialEq)]
pub enum Value {
  IntConstant(i32),
  Var(Identifier),
}

/// TACKY Value representation
#[derive(Debug, Clone, PartialEq)]
pub enum Instruction {
  Return(Value),
  UnaryMinus {
    src: Value,
    dst: Value,
  },
  Complement {
    src: Value,
    dst: Value,
  },
  LogicalNot {
    src: Value,
    dst: Value,
  },
  BinaryOp {
    op: ast::BinaryOp,
    lhs: Value,
    rhs: Value,
    dst: Value,
  },
  Jump(Identifier),
  JumpIfZero {
    cond: Value,
    target: Identifier,
  },
  JumpIfNotZero {
    cond: Value,
    target: Identifier,
  },
  Label(Identifier),
  Copy {
    src: Value,
    dst: Value,
  },
}

#[cfg(test)]
mod tests {
  use crate::{ast, shared_types::Identifier, tacky};
  use pretty_assertions::assert_eq;
  #[test]
  fn construct_tacky() {
    let var0 = tacky::Value::Var(Identifier::new("tmp0"));
    let var1 = tacky::Value::Var(Identifier::new("tmp1"));
    let c0 = tacky::Value::IntConstant(37);
    let comp = tacky::Instruction::Complement {
      src: c0.clone(),
      dst: var0.clone(),
    };
    let uminus = tacky::Instruction::UnaryMinus {
      src: c0.clone(),
      dst: var1.clone(),
    };
    let ret = tacky::Instruction::Return(var1.clone());
    let func = tacky::Function {
      name: Identifier::new("main"),
      body: vec![comp.clone(), uminus.clone(), ret.clone()],
    };
    assert_eq!(func.name, "main");
    assert_eq!(func.body.len(), 3);
    let prog = tacky::Program { function: func };
    assert_eq!(prog.function.name, "main");
    let _div_op = tacky::Instruction::BinaryOp {
      op: ast::BinaryOp::Divide,
      lhs: c0.clone(),
      rhs: var0.clone(),
      dst: var1.clone(),
    };
    let _plus_op = tacky::Instruction::BinaryOp {
      op: ast::BinaryOp::Plus,
      lhs: c0.clone(),
      rhs: var0.clone(),
      dst: var1.clone(),
    };
    let _minus_op = tacky::Instruction::BinaryOp {
      op: ast::BinaryOp::Subtract,
      lhs: c0.clone(),
      rhs: var0.clone(),
      dst: var1.clone(),
    };
    let _times_op = tacky::Instruction::BinaryOp {
      op: ast::BinaryOp::Multiply,
      lhs: c0.clone(),
      rhs: var0.clone(),
      dst: var1.clone(),
    };
    let _remainder_op = tacky::Instruction::BinaryOp {
      op: ast::BinaryOp::Remainder,
      lhs: c0.clone(),
      rhs: var0.clone(),
      dst: var1.clone(),
    };
    let _not = tacky::Instruction::LogicalNot {
      src: c0.clone(),
      dst: var0.clone(),
    };
    let label_name = Identifier::new("label1");
    let _label = tacky::Instruction::Label(label_name.clone());
    let _jmp = tacky::Instruction::Jump(label_name.clone());
    let _j0 = tacky::Instruction::JumpIfZero {
      cond: var1.clone(),
      target: label_name.clone(),
    };
    let _jn0 = tacky::Instruction::JumpIfNotZero {
      cond: c0.clone(),
      target: label_name.clone(),
    };
    let _cpy = tacky::Instruction::Copy {
      src: c0.clone(),
      dst: var0.clone(),
    };
  }
}
